### PR TITLE
benchmarking Podman: proof of concept

### PR DESCRIPTION
**Pull request by vrothberg**
Add a proof of concept for benchmarking Podman.  The benchmarks are
implemented by means of the end-to-end test suite but hidden behind
a `benchmarks` build tag.  Running `make localbenchmarks` will run
`test/e2e` with the specific build tag and set ginkgo's "focus" to
the specific "Podman Benchmark Suite" to only run this spec and skip
all others.

ginkgo will print a report before terminating listing the CPU and memory
stats for each benchmark.  New benchmarks can easily be added via the
`newBenchmark` function that also supports adding an `init()` function
to each benchmark which allows for performing certain setups for the
specific benchmark.  For instance, benchmarking `podman start` requires
creating a container beforehand.

Podman may be called more than once in the main function of a benchmark
but note that the displayed memory consumption is then a sum of all
Podman invocations.  The memory consumption is collected via
`/usr/bin/time`.

A benchmark's report is split into CPU and memory as displayed below:

```
[CPU] podman images:
  Fastest Time: 0.146s
  Slowest Time: 0.187s
  Average Time: 0.180s ± 0.015s
[MEM] podman images:
  Smallest: 41892.0KB
   Largest: 42792.0KB
   Average: 42380.7KB ± 286.4KB
```

Note that the benchmarks are not wired into the CI yet.  They are meant
as a proof of concept.  More benchmarks and the plumbing into CI will
happen in a later change.

Signed-off-by: Valentin Rothberg <vrothberg@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM

**Comment by vrothberg** [comment]
@baude PTAL

I consider the spike done. Once merged, I want to sit down with you and check which commands (and variations) we want to benchmark.

File: test/utils/utils.go [hunk 1]
```diff
@@ -44,6 +46,7 @@ var (
 	DefaultWaitTimeout   = 90
 	OSReleasePath        = "/etc/os-release"
 	ProcessOneCgroupPath = "/proc/1/cgroup"
+	RunBenchmarks        = false
```
Comment (vrothberg) [inline test/utils/utils.go:49]: This can be removed.

File: test/e2e/benchmarks_test.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	allBenchmarks []benchmark
+)
+
+// An internal struct for queuing benchmarks.
+type benchmark struct {
+	// The name of the benchmark.
+	name string
+	// The function to execute.
+	main func()
+	// Function is run before `function`.
```
Comment (vrothberg) [inline test/e2e/benchmarks_test.go:33]:
```suggestion
	// Function is run before `main`.
```

**Comment by vrothberg** [comment]
@baude PTAL

**Comment by TomSweeneyRedHat** [comment]
LGTM
@edsantiago PTAL

File: Makefile [hunk 3]
```diff
@@ -570,6 +570,14 @@ localintegration: test-binaries ginkgo
 .PHONY: remoteintegration
 remoteintegration: test-binaries ginkgo-remote
 
+.PHONY: localbenchmarks
+localbenchmarks: test-binaries
+	ACK_GINKGO_RC=true $(GOBIN)/ginkgo \
+		      -focus "Podman Benchmark Suite" \
+		      -tags "$(BUILDTAGS) benchmarks" -noColor \
+		      -noisySkippings=false -noisyPendings=false \
+		      test/e2e/. HACK=hack/.
```
Comment (edsantiago) [inline Makefile:579]: `HACK=hack/.` seems like a NOP here, very likely a copy-paste error. Possibly harmless, but misdirecting and time-wasting.

**Review by edsantiago — COMMENTED** [review]
Fails for me on my laptop:
```
Failed to compile e2e:

# github.com/containers/podman/v4/test/e2e [github.com/containers/podman/v4/test/e2e.test]
test/e2e/benchmarks_test.go:60:3: undefined: tempdir
test/e2e/benchmarks_test.go:64:33: undefined: tempdir
```
Not sure if this is a problem on my end, I will investigate but wanted to get this sent out early

File: test/e2e/benchmarks_test.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+var _ = Describe("Podman Benchmark Suite", func() {
+	var (
+		timedir    string
+		err        error
+		podmanTest *PodmanTestIntegration
+	)
+
+	setup := func() {
+		tempdir, err = CreateTempDirInTempDir()
```
Comment (edsantiago) [inline test/e2e/benchmarks_test.go:60]: Looks like the error is here, `=` should be `:=` and the `err` in line 55 needs to go.

But then, if I fix that, I somehow end up with a tmpdir called `/tmp/root` which is terrifying. Now I'm trying to figure out the source of that bug, but if it's easier for you to track down, I'd appreciate getting that fixed.

**Review by edsantiago — CHANGES_REQUESTED** [review]
Yeah, there's definitely something super-broken here wrt tempdir. Please fix in such a way that there is no use of `/tmp/root` or `/tmp/runroot` or anything like that; I'll take a fresh look once that's resolved.

**Comment by vrothberg** [comment]
> Yeah, there's definitely something super-broken here wrt tempdir. Please fix in such a way that there is no use of `/tmp/root` or `/tmp/runroot` or anything like that; I'll take a fresh look once that's resolved.

Great catch! Note that it predates this PR; it's using what other e2e tests are doing already.  I prefer to defer that to another PR.

**Review by edsantiago — APPROVED** [review]

**Comment by edsantiago** [comment]
>Note that it predates this PR;

Indeed it does - how can I not have caught this before...? Thanks for checking, and for fixing the other problems I noted. LGTM.

**Pull request merged**